src/Chapter3.hs

```diff
@@ -480,6 +504,38 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data City = City {
+  castle :: Castle, 
+  buildings :: Buildings, 
+  houses :: [House]
+}
+
+data Castle = None | Castle String | CastleWithWall String
+data Buildings = Church | Library
+data House = House {people :: People}
+data People = One | Two | Three | Four
+
+buildCastle :: City -> String -> City
+buildCastle city name = case castle city of
+  CastleWithWall _ -> city {castle = CastleWithWall name}
+  _ -> city {castle = Castle name}
+
+buildHouse :: City -> House -> City
+buildHouse city house = city {houses = house : houses city}
```

Comment:
According to our task, each house can contain from 1 to 4 people, But your `House` data type is a wrapper around `Int`, which means that users of this function can put any number of people inside the `House` (even the negative number). You can restrict this by introducing more checks.

This approach is usually called "smart constructor". To avoid building houses with more than 4 people you need to restrict usage of the Constructor and allow to create `House`s only with your `mkHouse` function.
You can read a bit more on the pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#smart-constructor

Generally, in here, you could restrict that on type level, by using something else than `Int` 🙂